Sources/HAP/Server/Configuration.swift

```diff
@@ -7,25 +7,48 @@ fileprivate let logger = getLogger("hap.device")
 func generateIdentifier() -> String {
     do {
         return Data(bytes: try Random.generate(byteCount: 6))
-            .map { String($0, radix: 16, uppercase: false) }
+            .map { String($0, radix: 16, uppercase: true) }
             .joined(separator: ":")
     } catch {
         fatalError("Could not generate identifier: \(error)")
     }
 }
 
+// Generate a valid random setup code, used to pair with the device
+func generateSetupCode() -> String {
+    let n1 = arc4random_uniform(1000)
+    let n2 = arc4random_uniform(100)
+    let n3 = arc4random_uniform(1000)
+    let setupCode = String(format: "%03ld-%02ld-%03ld", n1, n2, n3)
+    
+    if !Device.isValid(setupCode: setupCode) {
+        return generateSetupCode()
+    }
+    return setupCode
+}
+
+// Generate a random four character setup key, used in setupURI and setupHash
+func generateSetupKey() -> String {
+    return String(arc4random_uniform(1679616), radix:36, uppercase: true)
+}
+
+
 typealias PrivateKey = Data
 
 // The device maitains a configuration number during its life time, which
 // persists across restarts of the app.
 internal struct Configuration: Codable {
     let identifier: String
+    let setupCode: String
+    let setupKey: String
     let publicKey: PublicKey
     let privateKey: PrivateKey
 
     /// Initializes a new configuration
-    init() {
+    init(defaultSetupCode: String = "automatic") {
         identifier = generateIdentifier()
+        setupCode = defaultSetupCode != "automatic" ? defaultSetupCode : generateSetupCode()
+        setupKey = generateSetupKey()
```

Comment:
Why do we need this on the configuration? Do we need to restore both the code and key when restarting the device? Or can we regenerate them every time?

Reply:
I don't have access to the full spec, only what is made public, but I see in implementations such as Apples HomeKit Accessory Simulator and homebridge that these are preserved.